### PR TITLE
Improved cache basic structure

### PR DESCRIPTION
**Pull request by csviri**
For details see: https://github.com/java-operator-sdk/java-operator-sdk/issues/356

### CODE REVIEW

File: docs/.gitignore [hunk 1]
```diff
@@ -1,4 +1,4 @@
-.jekyll-cache/
+.jekyll-resourceCache/
```
Comment (metacosm) [inline docs/.gitignore:1]: I don't think that this change, or the next one, is intended…

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/CaffeinCacheAdaptor.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package io.javaoperatorsdk.operator.processing.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import io.fabric8.kubernetes.client.CustomResource;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+public class CaffeinCacheAdaptor implements ResourceCache {
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/CaffeinCacheAdaptor.java:9]: Should be `CaffeineCacheAdaptor`, imo.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/PassThroughResourceCache.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.javaoperatorsdk.operator.processing.KubernetesResourceUtils;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PassThroughResourceCache<R extends CustomResource> {
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/PassThroughResourceCache.java:17]: This should implement `ResourceCache`.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/PassThroughResourceCache.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  private static final Logger log = LoggerFactory.getLogger(PassThroughResourceCache.class);
+
+  private ResourceCache resourceCache;
+  private MixedOperation<R, KubernetesResourceList<R>, Resource<R>> client;
+  private final ObjectMapper objectMapper;
+  private Map<String, CustomResourceIdForClient> customResourceNameId = new ConcurrentHashMap<>();
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/PassThroughResourceCache.java:24]: Why do we need this?

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/PassThroughResourceCache.java:24]: @metacosm this is WIP for now, just did a draft release so it's easyer to see chnages. There is now a more fundamnetal issue, how we propagate UUID-s from event sources. Basically you cannot read CR from the server by UUID, so will have to  make there more changes also to EventSource structure.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/PassThroughResourceCache.java (hunk 3 above)
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/cache/PassThroughResourceCache.java:17]: Not sure if we want to go that way. If we want to handle caches efficiently, we always need the passthrough cache, and just the underlying store needs to be replaceable, that is why it does not implement ResourceCache.

File: docs/.gitignore (hunk 1 above)
Comment (csviri) [reply docs/.gitignore:1]: Haha, good point, will revert it.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/DefaultEventHandler.java [hunk 5]
```diff
@@ -64,19 +66,24 @@ public void handleEvent(Event event) {
...
   }
 
-  private void executeBufferedEvents(String customResourceUid) {
-    boolean newEventForResourceId = eventBuffer.containsEvents(customResourceUid);
-    boolean controllerUnderExecution = isControllerUnderExecution(customResourceUid);
+  private void executeBufferedEvents(CustomResourceID customResourceID) {
     Optional<CustomResource> latestCustomResource =
-        customResourceCache.getLatestResource(customResourceUid);
+        customResourceCache.getLatestResource(customResourceID);
+    if (latestCustomResource.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/DefaultEventHandler.java:78]: this is wip. pls don't comment on it yet. Will change the draft from to real PR if it's out of incubation.

**Pull request closed without merging**